jl4-lsp/src/LSP/L4/Actions.hs

```diff
@@ -96,6 +103,54 @@ gotoDefinition pos m positionMapping = do
   newRange <- toCurrentRange positionMapping lspRange
   pure (Location (fromNormalizedUri range.moduleUri) newRange)
 
+
+-- ----------------------------------------------------------------------------
+-- Ladder evalApp
+-- ----------------------------------------------------------------------------
+
+evalApp
+  :: MonadIO m
+  => TypeCheckResult
+  -> Ladder.EvalAppRequestParams
+  -> RecentlyVisualised
+  -> EL.Environment
+  -> ExceptT (TResponseError method) m Aeson.Value
+evalApp tcRes evalParams recentViz evalEnv =
+  case Ladder.lookupEvalAppMaker recentViz.vizState evalParams.appExpr of
+    Nothing -> defaultResponseError "No eval app directive maker found" -- TODO: Improve error codehere
+    Just evalAppMaker -> do
+      let evalAppDirective = evalAppMaker evalParams
+      (_, results) <- liftIO $ EL.execEvalModuleWithEnv evalEnv (evalAppDirective `prependToModule` tcRes.module')
+      Aeson.toJSON <$> getEvalResult results
+  where
+    prependToModule :: TopDecl Resolved -> Module Resolved -> Module Resolved
+    prependToModule newDecl (MkModule ann nuri (MkSection sann sresolved maka decls)) =
+      MkModule ann nuri (MkSection sann sresolved maka (newDecl : decls))
+
+    evalResultToLadderEvalAppResult :: Monad m => EL.EvalDirectiveResult -> ExceptT (TResponseError method) m EvalAppResult
+    evalResultToLadderEvalAppResult (EL.MkEvalDirectiveResult _ res) = case res of
+      Right (EL.MkNF val) -> 
+        case val of
+          EL.ValConstructor r [] -> EvalAppResult <$> toUBoolValue r
+          _                      -> throwExpectBoolResultError
+      Right EL.ToDeep    -> throwExpectBoolResultError
+      Left err           -> defaultResponseError $ Text.unlines $ EL.prettyEvalException err
+
+    throwExpectBoolResultError :: Monad m => ExceptT (TResponseError method) m a
+    throwExpectBoolResultError = defaultResponseError "Ladder visualizer is expecting a boolean result (and it should be impossible to have got a fn with a non-bool return type in the first place)"
+
+    toUBoolValue :: Monad m => Resolved -> ExceptT (TResponseError method) m Ladder.UBoolValue
+    toUBoolValue resolved = case getUnique resolved of
+      u | u == falseUnique -> pure Ladder.FalseV
+        | u == trueUnique  -> pure Ladder.TrueV
+        | otherwise        -> throwExpectBoolResultError
+
+    -- | Assumes that the order of the eval results is the same as the order of the eval directives.
```

Comment:
have you checked that this assumption upholds on modules with multiple eval directives?